consensus/istanbul/core/prepare_test.go

```diff
@@ -28,6 +28,62 @@ import (
 	"github.com/ethereum/go-ethereum/crypto/bls"
 )
 
+func TestHandlePreparedCertificate(t *testing.T) {
+	N := uint64(4) // replica 0 is the proposer, it will send messages to others
+	F := uint64(1)
+	sys := NewTestSystemWithBackend(N, F)
+	view := istanbul.View{
+		Round:    big.NewInt(0),
+		Sequence: big.NewInt(1),
+	}
+	proposal := makeBlock(0)
+
+	testCases := []struct {
+		certificate istanbul.PreparedCertificate
+		expectedErr error
+	}{
+		{
+			// Valid PREPARED certificate
+			sys.getPreparedCertificate(t, view, proposal),
+			nil,
+		},
+		{
+			// Invalid PREPARED certificate, duplicate message
+			func() istanbul.PreparedCertificate {
+				preparedCertificate := sys.getPreparedCertificate(t, view, proposal)
+				preparedCertificate.PrepareOrCommitMessages[1] = preparedCertificate.PrepareOrCommitMessages[0]
+				return preparedCertificate
+			}(),
+			errInvalidPreparedCertificateDuplicate,
+		},
+		{
+			// Invalid PREPARED certificate, hash mismatch
+			func() istanbul.PreparedCertificate {
+				preparedCertificate := sys.getPreparedCertificate(t, view, proposal)
+				preparedCertificate.PrepareOrCommitMessages[1] = preparedCertificate.PrepareOrCommitMessages[0]
+				preparedCertificate.Proposal = makeBlock(1)
+				return preparedCertificate
+			}(),
+			errInvalidPreparedCertificateDigestMismatch,
+		},
+		{
+			// Empty certificate
+			istanbul.EmptyPreparedCertificate(),
+			errInvalidPreparedCertificateNumMsgs,
+		},
+	}
+	for _, test := range testCases {
+		for _, backend := range sys.backends {
+			c := backend.engine.(*core)
+			err := c.handlePreparedCertificate(test.certificate)
+			if err != test.expectedErr {
+				t.Errorf("error mismatch: have %v, want %v", err, test.expectedErr)
+			}
+		}
+	}
+}
+
+// TODO(asa): Test with PREPARED certificate
```

Comment:
How hard would it be to add a test case for this?

Reply:
Apparently not too hard.